xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+			Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+		},
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu               sync.Mutex
+	balancerCurrent  balancer.Balancer
+	balancerPending  balancer.Balancer
+	scToSubBalancer  map[balancer.SubConn]balancer.Balancer
+	pendingState     balancer.State
+	currentLbIsReady bool
+
+	swapMu            sync.Mutex
+	balRecentlyClosed balancer.Balancer
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
+		return
+	}
+
+	// I want this to not intersplice with other updateState() calls and for
+	// this to happen atomically (i.e. if this update isn't atomic writing to
+	// pending state concurrently and sending wrong one out for one
+	// updateState() call).
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
+	gsb.swapMu.Lock()
+	gsb.balRecentlyClosed = nil
+	gsb.swapMu.Unlock()
+
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	if bal == gsb.balancerPending {
+		// Cache the pending state and picker if you don't need to send it at
+		// this instant (i.e. the LB policy is not ready) you can send it later
+		// on an event like current LB exits READY.
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLbIsReady {
+			gsb.swap()
+		}
+	} else {
+		gsb.currentLbIsReady = state.ConnectivityState == connectivity.Ready
+		// In the specific case that the current lb exits ready, and there is a
+		// pending lb, you can forward the pending LB's cached State up to
+		// ClientConn and swap the pending into the current. "Otherwise, the
+		// channel will keep using the old policy until...the old policy exits
+		// READY." - Java
+		if !gsb.currentLbIsReady && gsb.balancerPending != nil {
+			gsb.swap()
+		} else {
+			// Even if there is a pending balancer waiting to be gracefully
+			// switched to, still forward current balancer updates to the Client
+			// Conn. Ignoring state + picker from the current would cause
+			// undefined behavior/cause the system to behave incorrectly from
+			// the current LB policies perspective. Also, the current LB is
+			// still being used by grpc to choose SubConns per RPC, and thus
+			// should use the most updated form of the current balancer.
+			gsb.cc.UpdateState(state)
+		}
+	}
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	finished := make(chan struct{})
+	go func() {
+		gsb.swapMu.Lock()
+		defer gsb.swapMu.Unlock()
+		gsb.balRecentlyClosed = gsb.balancerCurrent
+		gsb.balancerCurrent.Close()
```

Comment:
Yeah, I remember thinking about this earlier (pre forked goroutine). I remember thinking...we're calling Close() inside holding gsb.mu.Lock()...and we call Close(), which is bound to call back inline into a function which grabs the lock and induces deadlock. I figured someone would say something about it previously, so is there a reason you thought if it now with swapMu() deadlock vs. previous where it would deadlock with any callback?

Reply:
1. this is still calling `Close()` while holding `gsb.mu`, because `swap()` blocks until this `Close` goroutine is done, and `mu` is Unlocked only after `swap()` returns. Right?
2. As long as `Close()` is called with some mu, and if it triggers a callback that requires the same mu, it will cause a deadlock (no matter it's `gsb.mu` or `gsb.swapMu`).

Reply:
Doug figured out a solution to this in our 1:1. Second mutex protecting swap operation and UpdateSubConn operation. Call balancer.Close() within this mutex in forked goroutine but not while holding main mutex. Still use main mutex to protect reads/writes. However, due to the atomicity of both events, now when UpdateSubConnState reads from scToSubBalancer, it will have a check to see if still in subConnMap, which will determine whether the balnacer has been closed or not.